### PR TITLE
Reject CoreML delegation for unsupported input dtypes

### LINKED ISSUE #11686: CoreML partitioner errors out with 8-bit integer types
### 🐛 Describe the bug

When partitioning a model with Core ML that contains int8 or uint8 tensors, the partitioner will error out with a KeyError. The backend should ideally either correctly handle it (if supported) or not partition it.

Repro:
```py
import torch

from executorch.backends.apple.coreml.partition import CoreMLPartitioner
from executorch.exir import to_edge_transform_and_lower
from executorch.extension.pybindings.portable_lib import _load_for_executorch_from_buffer

class Model(torch.nn.Module):
    def forward(self, x):
        return torch.abs(x)

model = Model()
inputs = (torch.randint(0, 255, (1,10)).to(torch.uint8),)
lowered = to_edge_transform_and_lower(
    torch.export.export(model, inputs),
    partitioner=[CoreMLPartitioner()],
).to_executorch()

et_model = _load_for_executorch_from_buffer(lowered.buffer)
et_model([*inputs])
```

Output:
```
File [~/miniconda3/envs/executorch/lib/python3.10/site-packages/coremltools/converters/mil/frontend/torch/exir_utils.py:60](http://localhost:8888/lab/tree/~/miniconda3/envs/executorch/lib/python3.10/site-packages/coremltools/converters/mil/frontend/torch/exir_utils.py#line=59), in _construct_ct_tensor_type_from_torch(name, tensor, symbol_name_to_ct_range_dim)
     55 def _construct_ct_tensor_type_from_torch(
     56     name: str,
     57     tensor: torch.Tensor,
     58     symbol_name_to_ct_range_dim: Dict[str, RangeDim],
     59 ) -> TensorType:
---> 60     coreml_dtype = TORCH_DTYPE_TO_MIL_DTYPE[tensor.dtype]
     61     if coreml_dtype == types.int16:
     62         coreml_dtype = types.int32

KeyError: torch.uint8
```

### Versions

Nightly

### PR DESCRIPTION
**Pull request by john-rocky**
### Summary

`CoreMLPartitioner` queries `coremltools.converters.mil.frontend.torch.is_torch_fx_node_supported`
to decide whether a node can be delegated, but that function only inspects the op
name — it does not validate input dtypes.  When a graph contains a node whose
inputs use a dtype outside `TORCH_DTYPE_TO_MIL_DTYPE` (e.g. `torch.uint8` /
`torch.int8`), the partitioner happily tags it and the failure only surfaces
later from deep inside coremltools as a `KeyError`, aborting the whole export.

This change adds an explicit dtype check in `should_override_support` so such
nodes fall back to the portable backend, and wraps the support query in
`try`/`except` as a defensive measure for any other case where the call might
raise instead of returning `False`.

Fixes #11686.

### Test plan

Added `test_unsupported_dtype_does_not_crash_partitioner` which lowers
`torch.abs` on a `uint8` input via `to_edge_transform_and_lower` and asserts
the op is not delegated.  Verified locally on macOS 15 / Python 3.10 /
coremltools 9.0:

```
$ python -m unittest -v executorch.backends.apple.coreml.test.test_coreml_partitioner.TestCoreMLPartitioner
Ran 9 tests in 104.215s

OK
```

Authored with Claude.

cc @metascroy

### CODE REVIEW

File: backends/apple/coreml/partition/coreml_partitioner.py [hunk 1]
```diff
@@ -75,6 +88,18 @@ def should_skip_op_for_delegation(self, node_target_name: str) -> bool:
         return False
 
     def should_override_support(self, node) -> bool:
+        # https://github.com/pytorch/executorch/issues/11686
```
Comment (metascroy) [inline backends/apple/coreml/partition/coreml_partitioner.py:133]: Can this be factored out into a utility like "node_dtype_is_supported" that gets called here rather than inlining the logic?

**Comment by metascroy** [comment]
Looks great @john-rocky! Left one nit, and running CI now.

File: backends/apple/coreml/partition/coreml_partitioner.py [hunk 2]
```diff
@@ -32,6 +32,35 @@
 logger = logging.getLogger(__name__)
 logger.setLevel(get_coreml_log_level(default_level=logging.INFO))
 
+# Mirrors coremltools' TORCH_DTYPE_TO_MIL_DTYPE.  Ops whose inputs use any
+# other dtype (e.g. torch.uint8) cannot be lowered, so the partitioner must
+# reject them.  See pytorch/executorch#11686.
+_COREML_SUPPORTED_INPUT_DTYPES = {
```
Comment (metascroy) [inline backends/apple/coreml/partition/coreml_partitioner.py]: @john-rocky this dtype list restriction for CoreML is only for the graph inputs, right?

int8/uint8 are allowed for internal nodes (e.g., static quantization), so I'm not sure the partitioner should be rejecting them, especially b/c they are used in quantization. Did you run the quantizer tests on this change?

It seems we need a function like:

```
def has_unsupported_dtype(node) -> bool:
```

that checks:

1. Is the node a graph input (not an internal node)
2. Then checks against the allowed input dtype list (can we import the list from coremltools rather than redefine it ourselves)?

It also looks like coremltools supports int8 input dtype on iOS26+? (https://github.com/apple/coremltools/blob/main/coremltools/converters/mil/input_types.py#L233) 


The macOS unit tests cannot be run on this PR until you rebase.

**Comment by john-rocky** [comment]
You were right on both counts — pushed a fix:

1. **Limited the dtype check to user inputs only**, using `graph_signature.input_specs` (`kind == USER_INPUT`) so lifted parameters / buffers — which are also placeholder nodes in the FX graph — no longer trip it.  Concretely this unbreaks the static-quantization paths your tests exercise: on the previous revision `backends/apple/coreml/test/test_torch_ops.py` had **6 failures + 4 errors** because the int8 `dequantize_affine` / `dequantize_codebook` weights were placeholders; on the new revision **all 11 tests pass**.
2. **Import the dtype set from coremltools** (`TORCH_DTYPE_TO_MIL_DTYPE`) with a fallback for older coremltools versions that don't export it.  That keeps us aligned automatically if/when iOS26 widens the input-dtype set with int8.
3. The original repro (uint8 user input → `torch.abs`) still falls back to portable; the existing `test_unsupported_dtype_does_not_crash_partitioner` continues to pass.

Also rebased on `upstream/main` so the macOS unit-tests CI can run.

Local verification:

```
$ python -m unittest backends.apple.coreml.test.test_torch_ops
Ran 11 tests in 4.257s
OK

$ python -m unittest backends.apple.coreml.test.test_coreml_partitioner
Ran 9 tests in 10.898s
OK
```

File: backends/apple/coreml/partition/coreml_partitioner.py [hunk 3]
```diff
@@ -33,6 +33,63 @@
...
 
+def _coreml_graph_input_dtypes() -> set:
+    """The dtypes coremltools can map for *graph inputs*.
+
+    Imported from coremltools so we don't drift if Apple updates the table
+    (e.g. iOS26 adds ``int8``).  Falls back to a hand-written set for older
+    coremltools versions that don't expose ``TORCH_DTYPE_TO_MIL_DTYPE``.
+    """
+    try:
+        from coremltools.converters.mil.frontend.torch.exir_utils import (
```
Comment (metascroy) [inline backends/apple/coreml/partition/coreml_partitioner.py]: ExecuTorch pins to specific versions of coremltools (currently 9.0), so we don't need a fallback path.

File: backends/apple/coreml/partition/coreml_partitioner.py [hunk 4]
```diff
@@ -33,6 +33,63 @@
...
+            torch.float16,
+            torch.float32,
+            torch.float64,
+            torch.int16,
+            torch.int32,
+            torch.int64,
+        }
+
+
+def _unsupported_graph_input_dtype(
```
Comment (metascroy) [inline backends/apple/coreml/partition/coreml_partitioner.py:50]: IIUC, I think this requires a little more thought to do what you want, and it might require a few passes.

In the original EP, we need to tag nodes whose inputs are graph inputs and have unsupported dtypes (I think that is what you're doing here).

But partitioning itself creates EPs, with their own "user" inputs. So we also need to tag nodes whose  inputs are untagged nodes with unsupported dtypes?

Imagine we have something like:

```
x = linear(x)
x = x.to(torch.int8)
x = torch.sum(x)
```

In this case the user input and output are fp32.

And for whatever reason, suppose the to op doesn't delegate to CoreML. Then we have:

CoreML         ET                       CoreML
 linear -> [to(torch.int8)] ->  sum

In this case, CoreML will receive an EP for just the sum op that has an int8 "input", even though in the original EP, all IO is float.

File: backends/apple/coreml/test/test_coreml_partitioner.py [hunk 5]
```diff
@@ -338,6 +338,35 @@ def forward(self, x):
                 torch.allclose(et_outputs, eager_outputs, atol=1e-02, rtol=1e-02)
             )
 
+    def test_unsupported_dtype_does_not_crash_partitioner(self):
```
Comment (metascroy) [inline backends/apple/coreml/test/test_coreml_partitioner.py]: Can we add many more tests here showing different cases of graph IO, intermediate, etc?

**Comment by john-rocky** [comment]
Pushed a follow-up addressing all three:

- **L44 fallback**: dropped.  Now imports `TORCH_DTYPE_TO_MIL_DTYPE` directly from coremltools.
- **L341 more tests**: added two cases on top of the existing one — a positive guard (fp32 user input is still delegated) and a mixed case (one fp32 user input + one uint8 user input → only the abs(uint8) branch falls back, the relu(fp32) branch is delegated).  All three pass locally; the existing `test_torch_ops.py` quantizer suite (11 dequantize_affine / dequantize_codebook tests) also still passes, which is what surfaced the user-input vs. lifted-parameter distinction in the first place.
- **L61 partition-boundary case**: you're right, and this version doesn't solve it.  This PR's check fires only on user-input placeholders, which catches the original #11686 repro but not the case where partitioning itself promotes an intermediate int8 edge into a sub-EP "user input".  Doing that correctly is what you said — multi-pass: first pass collects which nodes the partitioner is willing to delegate, second pass walks the resulting boundary edges and untags any node whose sub-EP input dtype isn't in the supported set, possibly iterating to a fixed point.  I'd rather land that as a follow-up because (a) it changes the partitioner's contract from "per-node decision" to "global decision", and (b) the failure mode it would prevent is fairly narrow — it requires a partition split at exactly an int8 edge that wasn't the user input.  I'm happy to write that follow-up if you want; want me to file it as a separate PR, or expand this one?

```
$ python -m unittest backends.apple.coreml.test.test_coreml_partitioner -v
Ran 11 tests in 12.4s
OK

$ python -m unittest backends.apple.coreml.test.test_torch_ops
Ran 11 tests in 5.688s
OK
```
